### ISSUE TITLE
Reads starting at position 1 are ignored when retrieving a region in samtools view

### ISSUE THREAD

**Issue by david-castillo** [samtools/samtools#1240]
#### Are you using the latest version of [samtools]

samtools 1.10
Using htslib 1.10.2
Copyright (C) 2019 Genome Research Ltd

 #### Please describe your environment.
Linux 4.15.0-88-generic
Ubuntu x64
installed from bioconda channel 

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

Hi,

I have a sample bam file that I've just created to illustrate the issue. It contains 7 paired reads that can be retrieved with:

`samtools view extract.bam`

SRR5077821.26763832	0	chrI	1	0	66P	=	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1
SRR5077821.26763826	1040	chrVI	1	0	66P	chrXVI	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1
SRR5077821.26763831	1040	chrVI	2	0	66P	chrXVI	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1
SRR5077821.26763827	1026	chrVI	2	0	66P	chrXVI	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1
SRR5077821.26763830	0	chrVI	65	0	66P	chrXVI	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1
SRR5077821.26763828	0	chrVI	66	0	66P	chrXVI	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1
SRR5077821.26763829	1026	chrVI	66	0	66P	chrXVI	813206	66	*	*	TC:i:1	S1:i:1	S2:i:1

If instead I want to extract a specific contig or region like:

`samtools view extract.bam chrI`

Then my reads mapped at position 1 are just ignored. I can't figure out why because it doesn't seem to depend on the contig or flag.

Am I doing something wrong?

Thanks

David

**Comment by jmarshall** [comment]
Your reads omit the SEQ and QUAL columns, using just the placeholders `*`. Because of this they are considered to be 0-length for indexing purposes, as they cover zero coordinate positions.

Read SRR5077821.26763831 thus extends from (1-based) position 2 up to but not including position 2. When you do a query for `samtools view extract.bam chrVI`, this read is printed because it is considered to overlap the requested 1…∞ region.

OTOH read SRR5077821.26763832 on chrI extends from (1-based) position 1 up to but not including position 1. This is **not** considered to overlap the requested 1…∞ region, so the read is not printed.

This is a bit of a corner case. In general, omitting the SEQ and QUAL columns is fairly unusual.

**Comment by david-castillo** [comment]
Thanks. I see now.
This is a pseudo-bam we generate to store pairwise interactions, so we didn't care much about the CIGAR, as far as it didn't affect the retrieval but of course we were not aware of the 0-based, 1-based dependence. We assumed BAM as 0-based.
So, in our case is it enough to specify an M in the CIGAR for all the reads?

**Comment by jmarshall** [comment]
This has nothing to do with 0-based vs. 1-based representations — I was just being clear that I was describing this using the same 1-based coordinates as appeared in your SAM example.

What would be appropriate to use in your CIGAR strings depends on what it is that you're trying to represent. Certainly using `M` would suffice to make these queries work, as it would mean some reference bases were being consumed by the alignment.

However I'm not really familiar with how the `P` operator is supposed to be used. Also there's a point of view that the tools should do something sensible even for reads in which the CIGAR represents something other than a standard mapping, and it would be possible to make samtools region queries work as expected for reads with `POS=1 CIGAR=66P` like these. See samtools/htslib#1063 for a(n incomplete) possible fix.

**Comment by david-castillo** [comment]
Thanks for the explanation, now it's clear to me why the reads are not appearing and that was the main point of my question.
I'll figure out what to do in our particular case and keep an eye on the possible modification of htslib.

Thank you again

**Comment by daviesrob** [comment]
I think this is definitely a misuse of the `P` operator, and `M` should be used instead.  However it is possible to get a similar situation with only inserted bases on alignments made by assemblers.  Consider this alignment where an existing 40 base ref starts `CCACC...` and `*` is a padding character:
```
Ref ********CCACACCACACCCACACACC********CACACACCACACCACACACC********
R1  ACGTACGTCCACACCACACCCACACACC
R2  ACGTACGT
R3          CCACACCACACCCACACACC
R4          CCACACCACACCCACACACCTGTGTCTC
R5                              TGTGTCTC
R6                              TGTGTCTCCACACACCACACCACACACC
R7                                      CACACACCACACCACACACC
R8                                      CACACACCACACCACACACCTGGCCGGT
R9                                                          TGGCCGGT
```

Here there are three reads which consume no reference bases - `R2` before the start, `R5`
 in the middle and `R9` at the end.  Despite not aligning to the reference they can be positioned by comparing to other reads with insertions.

Unfortunately the [SAM specification](https://samtools.github.io/hts-specs/SAMv1.pdf) doesn't say much about `POS` for reads that consume no reference bases. My guess is that it has to be the position after the insertion because going for the one before would break `R2` in BAM files (it would need to be -1, which is unmapped).  It also puts`R5` at the same position as `R6` which also has a leading insertion, but has the side-effect of placing `R9` after the end of the reference.  Possibly we need some hts-spec maintainer deliberation on this subject, but for now I'll carry on with `POS` as the reference base after the insertion.

So the question really is, what should we do with inserted bases at the start / end of ranges?  At the moment, `R2` is dropped by [this condition](https://github.com/samtools/htslib/blob/8859b09/hts.c#L3335) because it ends before the start of any specifiable range.  Special-casing the zero-length alignments would attach them to the range following the insertion, so `ref:1-20` would get `R1, R2, R3, R4` but not `R5`; and `ref:21-40` would get `R5, R6, R7, R8` but not `R9`.  `ref` would get all of them, but only because we set the end point to a very large number and not the actual reference length.  Neither range would give the full set of reads for the insertion between bases 20 and 21 - is this acceptable, or do we want to include them in both ranges, so you get to see the full picture?  It might not be too difficult to make the iterator go one base beyond the end point and return any reads with leading insertions, which would also make `R9` become visible for `ref:21-40`.  Likewise, `R4` could be included in the `ref:21-40` range by starting one base early, but only returning reads with trailing insertions.

So, which reads do we actually want to return for each range?

(Note that the issue with pads is pretty much the same as for insertions.  The problem is also a bit moot for mpileup as currently it has no way of returning a leading insertion partly due to limitations in its interface, and partly because it needs to know about reads like `R4` that may come before the region the pileup is being generated over.)

**Comment by jkbonfield** [comment]
I think you guess wrong from what I recall.

I had to jump through hoops to get this working in Gap5's pileup code.   The pileup code presents a list of all the sequences with data aligned to a specific point.  Insertions follow the previous base, so it's A and A+GT for a GT insertion.  If however you start on the GT you're hosed as you never see that in the pileup.  Given Gap5 was perfectly capable of handling such things outside of SAM/BAM, and use of P opcode to get multiple alignment within an insertion, it needed some careful trickery.

My "[level 10 hack](https://sourceforge.net/p/staden/code/HEAD/tree/staden/trunk/src/gap5/sam_pileup.c#l428)" in gap5 was to prepend all incoming CIGAR strings with 1D, decrement the position, and then in the pileup algorithm increment pos once more and strip off leading D operations.

Such a hack though probably isn't going to work for base 1, as that's pos 0 internally and decrementing it may not work depending on how we're doing our maths and range checking.

Basically this is a broken case in the file format itself that wasn't thought of at the time :/

**Comment by daviesrob** [comment]
I've attached my example in SAM form.

[ex.zip](https://github.com/samtools/samtools/files/4611166/ex.zip)

Gap5 does display it correctly; not so much luck with mpileup and tview due to the leading insertion issue.

![gap5_insertions](https://user-images.githubusercontent.com/3234562/81592952-a7658280-93b6-11ea-9e05-d7f6aaebbe57.png)

**Comment by jkbonfield** [comment]
Splendid!  Gap5's pileup code though is very different by design so hard to lift over the ideas. (What I could I've done already I think with the PR that added P op and bug-fixed the I/D combos.)

Specifically gap5's code generates a callback per column, with a co-ordinate of N+M for Mth base at ref position N, where M extends to the largest insertion.  Pileup is strictly one iteration per N and it is up to the caller to handle the insertion and how all the reads multiply align against each other,  however  it's obviously more focused on reference-based analysis so that may make sense.

I think however there really will be no fix for pileup other than a similar hack to the one I employed in Gap5, of preceeding all CIGARs with a ref-match operation (I chose 1D as it's easy to filter out leading deletions, but N would work too) and fudging the results downstream.  It'd break the interface though as the calling code is now exposed to an initial fake base.  There is no solution that I can think of that doesn't do that.

**Comment by jkbonfield** [comment]
> ```
>             1                 20       21                 30
>             |                  |        |                  |
> Ref ********CCACACCACACCCACACACC********CACACACCACACCACACACC********
> R1  ACGTACGTCCACACCACACCCACACACC
> R2  ACGTACGT
> R3          CCACACCACACCCACACACC
> R4          CCACACCACACCCACACACCTGTGTCTC
> R5                              TGTGTCTC
> R6                              TGTGTCTCCACACACCACACCACACACC
> R7                                      CACACACCACACCACACACC
> R8                                      CACACACCACACCACACACCTGGCCGGT
> R9                                                          TGGCCGGT
> ```

By definition a range query from 21 to 30 should bring back only data with a base that aligns to reference bases 21 to 30.  That means not R5 or R9, and obviously not R4 as none of them have any base mapping to that range.

However the same logic means a query from 1 to 30 also shouldn't return R2 nor R9.

So how do we get those reads?  I think we have to relax our range a bit.  If we want reads that align *only* to insertions before the first base and ones that align only to insertions beyond the end of the reference then we should ask for range 0 to 31.  I think that is the logical fix here, simply to permit broader ranges in order to obtain those.

The case of how to handle zero length reads is perhaps a bit different.
